Issue #1645 Interface Support Mock Testing

kccarter76 opened the pull request:
<!-- markdownlint-disable MD002 MD041 -->

### Issues

Feature enhancement resolves #1645 

### Description

added the ability to use an interface IODataQueryOptions<TEntity> as a parameter on a controller method.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

feature needs to still be documented and the issue #1645 updated with Docs Needed.

bitsforthought committed:
Adding E2E test coverage for $select support of complex properties

In the spirit of TDD, I wrote a failing test before attempting to fix this
issue: https://github.com/OData/WebApi/issues/196

Unfortunately, the fix looks more like a large and complex feature than just
a simple bug fix, and I don't have bandwidth to take it on right now.
However, the test may still be useful for the next person to make the
attempt.

biaol-odata committed:
Issue #170 --- Verify SingleResultWithEmptyIQueryable

bitsforthought committed:
Splitting SelectExpand E2E tests into two separate classes

While trying to add complex types to the test model for SelectExpandTests, I
ran into an issue with EF where some existing tests failed due to EF's own EDM
model validation. This was happening because it assumes the new complex type
is actually an entity type, even though it doesn't have a key.

My solution is just to factor out the EF-related tests into their own class,
SelectExpandEFTests. This should allow the addition of complex types to the
SelectExpandTests test model without breaking any of the existing tests.

bitsforthought committed:
Removing whitespace per PR feedback

bitsforthought committed:
Replacing property with field

bitsforthought committed:
Merge PR#1458: Allow recursive comple3x types in OData model builder

xuzhg committed:
Fix the policheck problem

xuzhg committed:
add a classic sample project

xuzhg committed:
fix the BaseAddressFactory in ODataMediaTypeFormatter

xuzhg committed:
fix issue #1471 about the non-odata routing with DataContact & DataMember

xuzhg committed:
Change the OptionalReturn to ReturnNullable in OperationConfiguration

xuzhg committed:
Add the optional parameter into the model builder

xuzhg committed:
Fix #1387: add instance for the select all binder for perf

xuzhg committed:
change the property and fluent API in parameter configuration

xuzhg committed:
Return Task<> from method of controller doesn't pick up odata output formatter

biaol-odata committed:
Add support for Json Batch.

xuzhg committed:
issue #1434: add the OData-Version into the No-Content response

biaol-odata committed:
Update ODL dependencies (#1508)

* Update ODL dependencies to 7.5.0
    - Update package versions.
    - Revert test workaround for the pending 7.5.0 release.
    - Disable obsolete warnings for AggregateTransformationNode.Expressions introduced in v7.5.0

* Update dependencies on classic side and performance.sln.

xuzhg committed:
move  &  execute as late as possible

athoscouto committed:
Entity set aggregation (#906)

* Adding support to entity set aggregations.

xuzhg committed:
update the version to 7.0 RTM (#1512)

AlanWong-MS committed:
[Feature] IN operator in WebAPI (#1497)

* Enable IN operator for $filter. Supported collections include paren-defined collections, property collections, and navigation properties.
* Added test cases to accompany change.

xuzhg committed:
update the version to 7.0.1.beta

xuzhg committed:
downgrade the newtonsoft.json version to 6.0.4 for classic (#1519)

* downgrade the newtonsoft.json version to 6.0.4 for classic

* change the newtonsoft.json version in classic test

xuzhg committed:
make covention and attribute routing working on optional parameter

xuzhg committed:
Add UseOData() extension methods for IApplicationBuilder

xuzhg committed:
Modify the public api

xuzhg committed:
change the variable name

xuzhg committed:
Change the set as protected set in ParameterConfiguration

biaol-odata committed:
Fix json batch response content type.

xuzhg committed:
Fix issue #1529: ETag missing on derived entity set

xuzhg committed:
Resolve the comments

xuzhg committed:
fix a typo

xuzhg committed:
update version to 7.0.1 RTM

xuzhg committed:
update the version to 7.0.2 beta

biaol-odata committed:
For netcore implementation, skip content serialization content for empty content.

xuzhg committed:
Fix the issue #850, (de)serialize using the Edm Enum member string

xuzhg committed:
change the comments

xuzhg committed:
EnumMember also accept the Model alias

xuzhg committed:
Resolve the comments and add more unit tests

xuzhg committed:
Fix issue #1555: ETagMessageHandler on controller will throw exception

andrewlock committed:
Don't assume port 80 when generating nextLink in ASP.NET Core #1559 (#1560)

* Don't assume port 80 when generating nextLink in ASP.NET Core #1559

If the port isn't set on `request.Host`, port 80 was assumed previously.
If the app is running behind a reverse proxy, and the port was not forwarded,
this can result in invalid URLs - for example an https URL using port 80
Instead, we just omit the port completely if it's not explictly set.

* Add unit tests for GetNextPageLink

* Add copyright notice

xuzhg committed:
Fix =* not returning dynamic properties

xuzhg committed:
resolve the comments

smariussorin committed:
Null check on ODataQueryParameterBindingAttribute for Net Core (#1588)

- Null check on ODataQueryParameterBindingAttribute for Net Core

bitsforthought committed:
CSDL returned from $metadata endpoint has incorrect "Unicode" attributes

This is a regression from 5.x to 6.x/7.x. It looks like the original fix was
never merged to the 6.0 development branch.

Issue: https://github.com/OData/WebApi/issues/736

xuzhg committed:
Add the iconUrl to the nuget spec

rasoulian committed:
Fix compile error in sample project (#1615)

- Add comma before MyLevel property.

kosinsky committed:
$filter=Nav/any(t:false) converted to Nav.any(_ => False) instead of Nav.any()

bitsforthought committed:
Improving the error message for when an unknown property is deserialized

This is addresses the unhelpful error message described in this GitHub issue:

https://github.com/OData/WebApi/issues/1421

bitsforthought committed:
Addressing PR feedback

KanishManuja-MS committed:
Merge pull request #1621 from brjohnstmsft/better-error-message

Improving the error message for when an unknown property is deserialized

kosinsky committed:
Remove contradiction

KanishManuja-MS committed:
Allow adding new members to a collection through a POST request (#1628)

* Enable adding members to collection of complex types

* Fix deserialization for primitive types

The ODataPrimitiveDeserializer did not support non-trivial (that cannot be cast directly from the string representation) primitive types like Binary, GUIDS, Spatial etc.

* Support cast uri as well

* Pass the type reference to ODL to parse correctly

* Do not use ConvertPrimitiveValue, add unit test for ODataPrimitiveDeserializer

* Fix unit tests

* Add unit tests for propertyrouting changes

* Undo unnecessary changes

* Add E2E test for PostToCollections

Add E2E test for PostToCollections

* typo

* E2E case for collection of Enums

* Keep original and add another testcase

* Account for the additional property on the test cases

* Add test cases for Binary and Guids

* Make the private helper functions for Enum deserialization tests accessible for other classes

* New E2E cases for posting to collection of complex type with inheritance

* Add more test cases for trivial primitive values

* Add a new date time object to a collection property

* Do not imply unique identifiers

* Change the response from Created to Ok to avoid exception, response is yet to be discussed

* fix spacing errors

* Fix Unit test

* Remove test case for the time being because of difference in behavior locally and on server

* Address self reviewed issues

* Improve the enum test case

* Address Sam's comments on the PR

* remove extra whitespace

* Add testcase to return representation

* Add using to Netcore as well

* Address more PR comments

* Address last of the PR comments

xuzhg committed:
Fix the security vulnerabilities in project dependencies:
1. Update the Microsoft.Data.OData vulnerability to 5.8.4
2. Update the Microsoft.AspNetCore.All to 2.0.9
3. Update the Microsoft.AspNetCore.Server.Kestrel.* to 2.0.4

biaol-odata committed:
Allow ODataFeature.TotalCount set to null. (#1667)

* Allow ODataFeature.TotalCount set to null.

* CR update: set this.totalCountSet = value.HasValue.

KanishManuja-MS committed:
Respect preference header for paging (#1673)

* Respect preference header for specifying maxpagesize

* Don't set pagesize in the enable query code; instead wait till options to maintain parity

* Fix single get request issue

* Revert "Merge branch 'Respect-preference-header-for-paging' of https://github.com/KanishManuja-MS/WebApi into Respect-preference-header-for-paging"

This reverts commit 14a45de5748e55c5c883c0ba24006e0c0cec6095, reversing
changes made to 338d857cfd40db58fc8cfb4db0d41eece9746f94.

* Update src/Microsoft.AspNet.OData.Shared/RequestPreferenceHelpers.cs

* Fix for classic version

* fix aspnet oddities

* Update src/Microsoft.AspNet.OData.Shared/RequestPreferenceHelpers.cs

* Use String instead of string to avoid performance build error

* Address PR comments

* Update src/Microsoft.AspNet.OData.Shared/RequestPreferenceHelpers.cs

* Fix spacing

freeranger committed:
Fixes #1600 (#1601)

HRasch committed:
Take and Top query use wrong Provider.CreateQuery Method #1617 (#1618)

* Take and Top query use wrong Provider.CreateQuery Method #1617

* Bugfix: Cast to IProviderInterface before scanning for the methods

* Based on review feedback, method searching is now cached

biaol-odata committed:
Individual query options to utilize DI config (#1664)

* For individual query options, pass in the container so that global config is available.

* CR comments.

KanishManuja-MS committed:
Fix typo in ODataApplicationBuilder.cs (#1681)

* Update src/Microsoft.AspNetCore.OData/Extensions/ODataApplicationBuilderExtensions.cs

* Update public API bsl

xuzhg committed:
Fix issue #1631, Singleton and single navigation property possible without key defined on its type (#1641)

RamjotSingh committed:
Fix #1591 -  Honor namespace value when registering Types using Builder (#1592)

* Fix for #1591 - OData Builder should honor assigned Namespace

* Fixing tests

* Enabling build on machines using VS2017 alone also fixing build for Preview version of VS

* Using String instead of string in expression (Build fix)

* Adding registration for Enum Types as well

* Updating comments

* Taking PR feedback

* Taking PR comments. Adding code to revert namespace to Default, if user sets it to null or empty

xuzhg committed:
Allowing the definition of partner relationships (#1685)

* Allowing the definition of partner relationships

* Change the C# style to make the VS2015 build pass

biaol-odata committed:
version bump to 7.1 RTM. (#1686)

biaol-odata committed:
update version to 7.1.1.beta

xuzhg committed:
Fix issue #1725 and related: make  on DateTimeOffset with Date work

kosinsky committed:
Ensure that all expand item options preserved when translating SELECT *

When WebAPI executed `$expand` and `$select` it takes SelectExpandClause and transforms it to apply all selects, level and expand filters. For example
it you have `WorkItems?$expand=Children` it will be automatically (in same cases) converted into
`WorkItems?$select=<list of all columns>&$expand=Children($select=<list of all columns>)`.
It's implemented as creating new SelectExpandClause, however other options inside `$expand` were ignored as a result `WorkItems?$expand=Children($filter=State eq 'Active')` was translated to `WorkItems?$select=<list of all columns>&$expand=Children($select=<list of all columns>)`

Fixing that issue

xuzhg committed:
Update README.md

robertmclaws committed:
[Regression] Fixes and coverage for primitives in URL parameters. (#1742)

* Fixes for regression described in #1741

- Optimizes the codepath so that primitives can be tested without a DI container.
- Adds the missed conversion for ConstantNodes.

- [X] Includes unit tests

* Update ODataModelBinderConverterTests.cs

Documentation adjustments.

raheph committed:
Merge pull request #1715 from kosinsky/users/kokosins/AnyFalse

Fix for Nav/any(f:false) conversion

xuzhg committed:
Update the Microsoft.AspNetCore.All version to meet security vulnerablilities

Commit:
Test and fix for issue 1802 (ODataEnumDeserializer.ReadInline fails with InvalidCastException when edmType comes from CsdlParser)

commit 0a383f4f6d03c17efb9658f189c4703877ec6b12
Author: David Jackman <djackman@microsoft.com>
Date:   Wed May 8 10:49:57 2019 -0600

    Change C# 7 syntax to C# 6 compatible so StyleCop will not think it's a syntax error.

commit df18e376b08b9302f985bab1be561dc5eae255c8
Author: David Jackman <djackman@microsoft.com>
Date:   Tue May 7 16:40:25 2019 -0600

    Test and fix for issue 1802 (ODataEnumDeserializer.ReadInline fails with InvalidCastException when edmType comes from CsdlParser)

kosinsky committed:
PR # 1793: Fix ODataModelBinderConverterTests in debug mode

xuzhg committed:
Update to OData Library v7.6, fix the failing test cases (#1825)

* Update to OData Library v7.6, fix the failing test cases

1. Navigation property binding sorted by the path string
2. Context Uri changes for $expand by add parathesis
3. Complex type parameter supports using ODataResourceValue
4. etc.

* remove unused changes

* Update the AspNetODataSample.Web

* Change the namespace of EntitySetTest

xuzhg committed:
Change the test case CanConfigureSingleProperty_MultipleBindingPath_For_NavigationProperties_WithComplex_Multiple for stable

KanishManuja-MS committed:
Custom NextLink generation with default SkipToken implementation (#1724)

* SkipToken implementation with nested nextlink

Update Microsoft.AspNet.OData.PublicApi.bsl

* make constructor and properties internal

* update public api

* Address PR comments

* Remove unnecessary lines

* Address PR comments

* Address final PR comments

* Fix perf build issues

* Address PR comments

kosinsky committed:
Support leading and trailing spaces in query options (#1823)

* Support leading and trailing spaces for query options

To be more tolerant for simple typos

* More samples and better formating

kosinsky committed:
Support $apply for EF Core (#1728)

* Ef Core InMemory provider works with aggregations

* $apply works with EF Core SQL provider

EF Core and EF6 need a bit different approach for aggregating IGrouping<>

* Reflect changes in BinderTests

* Comments and more EF Classic tests

* Stylecop issues

* Reaction on feedback

* Block Linq2SQL

* Added tests for enum property

* Fix typo

xuzhg committed:
Enable dollar ref after dollar expand

genusP committed:
fix #194. Support count in orderby and filter query options. (#1844)

Thanks for the contribution and quick follow-up!

joshcomley committed:
Fixing N+1 query bug on expands (#1653)

* Fixing N+1 query bug on expands

* Add Setting for CorrelatedSubqueryOptimization

mikepizzo committed:
401 support (#1839)

* Support OData 4.01

* Added Tests

gunars-skangals committed:
Fix assignment of CollectionExpandedProperty.TotalCount when using EF (#1693)

* PropertyContainer: convert result of TotalCount to long? before assignment

* SelectExpandNestedDollarCountTest: remove redundant argument

genusP committed:
Support default value for primitive and enum property (#1851)

* Support default value for primitive and enum property

* fix style issue

adzhiljano committed:
Fix SelectExpandBinder to take into account constant parameterization when using ModelId as part of a query(#1541) (#1542)

When ModelId is used as part of a query in the SelectExpandBinder, it should respect the ODataQuerySettings.EnableConstantParameterization property. Otherwise, there is no way to force a generation of a SQL parameter in the resulting query. Which can result in polluting the SQL Server query plan cache with unwanted query plans (containing constants).

genusP committed:
Improvement support continue-on-error in batch request (#1849)

hfaran committed:
Fix error response in ValidateODataBatchRequest (#1854)

CI failure was a build machine issue; @KanishManuja-MS re-ran and validated that all checks passed.  Merging this PR - thanks!

adzhiljano committed:
Fix EnableQuery attribute to take into account constant parameterization when PageSize is set (#1523) (#1535)

* Fix EnableQuery to take into account constant parameterization when PageSize is set (#1523)

When PageSize is being handled in the EnableQuery attribute it should respect the ODataQuerySettings.EnableConstantParameterization property. Otherwise, there is no way to force a generation of a SQL parameter in the resulting query. Which can result in polluting the SQL Server query plan cache with unwanted query plans (containing constants).

* Fix build and resolve conflicts

Galad committed:
Fix ordering the keys according to the ColumnAttribute.Order property (#1762)

KanishManuja-MS committed:
Remove constructor from public surface; upcoming PR will remove it (#1868)

kosinsky committed:
Flatten properties before using in aggregation to avoid nested queries. (#1855)

* Ef Core InMemory provider works with aggregations

* $apply works with EF Core SQL provider

EF Core and EF6 need a bit different approach for aggregating IGrouping<>

* Reflect changes in BinderTests

* Comments and more EF Classic tests

* Stylecop issues

* Reaction on feedback

* Block Linq2SQL

* Flatten properties before using in aggregation to avoid nested queries.

* Fix ordering of flattened properties

* Fix EntitySet aggregations

* Ensure that we checking SQL for E2E aggregation tests

* Stylecop issues

KanishManuja-MS committed:
Revert "Fix ordering the keys according to the ColumnAttribute.Order property (#1762)" (#1871)

This reverts commit 9519a42ae78217ccdf3dfb32a9fdf84c0e30d02e.

KanishManuja-MS committed:
Fix ordering the keys according to the ColumnAttribute.Order property (#1867)

Update ColumnAttributeEdmPropertyConvention.cs

KanishManuja-MS committed:
Update WebStack.versions.settings.targets (#1869)

KanishManuja-MS committed:
Update minimum version of ODL to be 7.6 (#1876)

* Update minimum version of ODL to be 7.6

* Add release note links to nuget.org

KanishManuja-MS committed:
Update the version number for the fix release

mikepizzo committed:
Include EnableCorrelatedSubqueryBuffering when copying from existing ODataQuerySettings. (#1889)

folkcoder committed:
Ensure generated skip token respects casing conventions (#1895)

* create skip token using edm rather than clr property name, and clarify variable names

* add end-to-end tests to ensure generated skip tokens match the camel-cased edm property name

Commit:
create test and fix issue

Commit:
replace var to PropertyInfo

mikepizzo committed:
create test and fix issue (#1907)

mikepizzo committed:
Merge pull request #1884 from kosinsky/features/PreserveExpandOptions

Ensure that all expand item options preserved when translating SELECT *

xuzhg committed:
Change AssemblyVersionAttribute without revision

KanishManuja-MS committed:
Support navigation properties on complex types (#1738)

* Enable expand on navigation properties on complex types

* Fix PR issues; Do not introduce any public elements

* fix build issues

* Update Microsoft.Test.E2E.AspNetCore.OData.csproj

* Add test case for open type

* Update SelectExpandNode.cs

* Update SelectExpandNode.cs

kosinsky committed:
Added more logging for flacky AggregateNavigationPropertyWorks test (#1916)

xuzhg committed:
Rename a country name for make Policheck happy

xuzhg committed:
Update to OData Library v 7.6.1, fix the failing test cases (#1919)

* Update to OData Library v 7.6.1, fix the failing test cases

* Change the Package dependency to 7.6.0 to 7.6.1

xuzhg committed:
Update the Performance project dependency to 7.6.1

xuzhg committed:
fix issue #1814: Add Entity property for Created and Updated Result

xuzhg committed:
Update the PublicApi base

adzhilyanov committed:
Support groupby with property path length greater than 2

KanishManuja-MS committed:
Add a compatibility flag for generating next link with negative top v… (#1920)

* Add a compatibility flag for generating next link with negative top values as well

* Fix failing test cases

* Update Public API and rename ODataCompatibilityOptions to CompatibilityOptions

KanishManuja-MS committed:
Recreate the DataBase if the Model is detected to be changed. (#1927)

Danielku15 committed:
Allow IN operator for enums (#1910)

* Ensure enum values are unwrapped for in operator

* fix enum in IN operator

KanishManuja-MS committed:
Update WebStack.versions.settings.targets (#1928)

xuzhg committed:
Enable $select improvement with path and nested query options (#1941)

* Enable $select improvement with select path and nested query options

1. Add the test case to verify the same derived property name
2. Fix the $ref test cases
3. Fix the base address case sensitive problem in the build agent.
4. Fix some failing e2e test cases
5.  Address the comments

habbes committed:
Fix #1965: Pass keyId parameter to controller actions of single-key entities

xuzhg committed:
Update the release version to 7.2.3

xuzhg committed:
Enable skip test case because the issue #196 is fixed

xuzhg committed:
Fix the byte[] and char[] property in select

odero committed:
Update readme

xuzhg committed:
Enable IRouter-based routing for ASP.NET Core 3.x

YishaiHornbacher committed:
Fix error message to state that value must be 1 or greater, instead of zero or greater.

xuzhg committed:
Update Nuget.exe and build.ps1

xuzhg committed:
Fix #1992, Missing Null check in SelectExpandNode class

xuzhg committed:
Update to .NET Core 3.1

xuzhg committed:
Update to release version 7.3.0 RTM

xuzhg committed:
Add collection navigation property into the sample

xuzhg committed:
Add the Lazy-loading into the .NET Core 3.1 sample project

xuzhg committed:
Fix the ToInt16 overflow during build

xuzhg committed:
Add a collection complex into sampel project

gathogojr committed:
Fix issue #1610: OData AddLink (CreateRef) error when performing batch operations, PR #2012

xuzhg committed:
Update Readme.md to use the latest build status badge

CarlosMontagut committed:
PR #2014: Fix performance drop when using a big model

mikepizzo committed:
Batch Fixes:
1) Fix batch dispose issue (error writing to a closed stream)
2) Fix async batch issues (synchronous write to an asynchronous reader error)

mikepizzo committed:
Fix multiple batching issues:
-Case-insensitive header comparison
-Default to JSON Batch response if request is JSON
-Make sure stream is released before reading next request within a batch
-Added/fixed tests for standard and .net core

mikepizzo committed:
Update WriteResponseAsync to be backward compatible with current synchronous writing behavior to avoid possible (though unlikely) breaking change.
Note that any types directly deriving from ODataBatchResponseItem would have to be recompiled due to changes in abstract base class (abstract method now takes a writeAsync parameter).  Seems highly unlikely that other custom types are deriving directly from ODataBatchResponseItem.
In .NET 5, we should clean up the overloads that write to the stream synchronously, and probably make most of this internal (no reason to make the WriteResponseAsync/WriteMessageAsync on ODataBatchResponseItem public).

xuzhg committed:
Fix the metadata level in ETag E2E test

xuzhg committed:
Change ETagsOtherTypesTest for independent parallel running

mikepizzo committed:
Merge pull request #2048 from mikepizzo/BatchFixes

Batch Fixes

xuzhg committed:
Fix DollarId test parallel running

Sreejithpin committed:
Merge pull request #2065 from xuzhg/FixDollarIdParallelTests

Fix DollarId test parallel running

xuzhg committed:
Fix MixScenarioTest and Inheritance parallel running problem

kosinsky committed:
Added support for compute() transformation in $apply (#1877)

* Added support for functions and constants in aggregate expression

* Tests for more cannonical functions

* Time tests

* Support for compute transformation in $apply

* Support sorting for columns introduced in compute

* Remove redundant code

* Fix flakyness in AggregateNavigationPropertyWorks

* Fix the ToInt16 overflow during build

* Fix typos

* Make _values field private

* Explicit way to mutate element type and lambda param

* Test to demonstrate that we could run compute on complex entity properties

* Support for complex types

* Simplify keepoing ElementType and LambdaParameter in sync

Co-authored-by: Sam Xu <saxu@microsoft.com>

Sreejithpin committed:
Fix issue#1962: Filtering on a property of casted object fails. PR #1986

kusc-leica committed:
Ensures nullable enums are properly handled for In operator. Issue #1676, PR#1933

xuzhg committed:
[Endpoint routing review]First draft to enable endpoint routing using dynamic controller (#2035)

* Enable endpoint routing using dynamic controller

* Update to 7.4.0

* Update the PublicAPI test cases

* Add the E2E test cases

* Change the startup to test the per-route configuration

* Some test codes about policy

* Use MatcherPolicy Instead of EndpointSelector to resolve the ambiguity

* Address the comments and add LinkGenerator test cases

* Revert the EF Core version to 2.2.0 to make Aggregation E2E test pass

kosinsky committed:
Ensure =compute() works with EFCore3.1, PR #2069

Commit:
Fix #1953: Generation of Skiptoken value with DateTime

Commit:
Fix #1947: Use ODataValueProvider, PR #1948

Commit:
Remove CaseInsensitiveResolver in test, PR #1946

xuzhg committed:
Enable $batch request with Endpoint routing

habbes committed:
Filter batch request headers copied to individual requests (#2064)

* Add test to verify batch handler does not copy all headers to individual requests

* Control which headers are copied from batch to individual requests

* Handle quoted and escaped preference header values in batch requests

* Fix regression in default batch handler tests

* Minor refactor of header handling in batch handler

* Use only comma as separator for preferences in Prefer header

* Handle multiple prefer headers and duplicate preferences in batch request

* Minor style fixes

* Refactor handling of batch headers

* Remove unnecessary Trim() in preference header handling

* Remove space from preference header separator

xuzhg committed:
Add the parameter name when create the parameter expression

xuzhg committed:
Revert the default null propagation to false for EF Core

tomimon committed:
PR #2067: Fix the build.ps1 in internaltion system environment

xuzhg committed:
Update the version to beta2

KanishManuja-MS committed:
Add support for derived type constraint annotation (#1977)

* Everything except properties.

Fix tests

Address PR comments

Remove Inline change due to a bug

* Address PR comments and take sam's input

parity between actions and functions

self review

Update tests

fix final set of issues

* Address PR comments

address pr comments

update method

xuzhg committed:
update the release version to 7.4 RTM

xuzhg committed:
Fix #2146 and #2141: the context Uri and location wrong with PathBase setting

xuzhg committed:
Update README.md

mikepizzo committed:
Fix issue #2009: sync calls within async routines

crgolden committed:
OData/WebApi#1837: SetDefaultODataOptions doesn't work/has dead code

xuzhg committed:
Fix the failing batch async test cases

gathogojr committed:
Fix logic for patching structured type with complex type dynamic property (#2160)

* Fix logic for patching entity with complex type dynamic property

* Support patching of complex type dynamic property in a structured type plus tests

* Use string.IsNullOrWhiteSpace

Co-authored-by: John Gathogo <jogathog@microsoft.com>

xuzhg committed:
update the version 7.4.1

SiberaIndustries committed:
Fix client-side evaluation for string comparison queries, PR #2053.

KanishManuja-MS committed:
Copy query settings correctly, PR #2169

habbes committed:
Fix EFCore memory lean when query contains collection constant, PR 2172

xuzhg committed:
Add System.Diagnostics.EventLog

Add System.Diagnostics.EventLog into ASP.NET Core 3.x E2E test project

kccarter76 committed:
#1645 okay this is the easy part, now I get to hunt down where the injection occurs

kccarter76 committed:
Create dotnet-core.yml

kccarter76 committed:
Update dotnet-core.yml

kccarter76 committed:
Update dotnet-core.yml

kccarter76 committed:
Update dotnet-core.yml

kccarter76 committed:
Update dotnet-core.yml

kccarter76 committed:
Update dotnet-core.yml

kccarter76 committed:
trying to get the github integration server building this code

kccarter76 committed:
#1645

kccarter76 committed:
linux is very picky about names

kccarter76 committed:
Update dotnet-core.yml

kccarter76 committed:
Rename ProductsCSDL.xml to ProductsCsdl.xml

kccarter76 committed:
no message

kccarter76 committed:
local test run completes succesful

kccarter76 committed:
Merge pull request #5 from kccarter76/dev/issue-1645

Dev/issue 1645

kccarter76 committed:
#1645 unit test added and all unit tests passing

kccarter76 committed:
Merge pull request #6 from kccarter76/dev/issue-1645

#1645 unit test added and all unit tests passing

kccarter76 committed:
Delete dotnet-core.yml

kccarter76 committed:
remove the code that the designer did not

kccarter76 committed:
#1645 moved the interfaces to the shared project and adjusted things base on vs studio.

kccarter76 committed:
#1645 updated the EnableQueryAttribute for AspNet

kccarter76 commented:
Base line file test\UnitTest\Microsoft.AspNet.OData.Test\PublicApi\Microsoft.AspNet.OData.PublicApi.bsl and output file D:\a\1\s\bin\Release\UnitTest\AspNet\Microsoft.AspNet.OData.PublicApi.out do not match, please check.
To update the baseline, please run:

copy /y "D:\a\1\s\bin\Release\UnitTest\AspNet\Microsoft.AspNet.OData.PublicApi.out" "test\UnitTest\Microsoft.AspNet.OData.Test\PublicApi\Microsoft.AspNet.OData.PublicApi.bsl"

I am going to need some help on your side to address this issue.

kccarter76 committed:
#1645 finally was able to generate the bsl

YoannaAtanasova committed:
Issue 2177 - Fix skip token encoding (#2197)

* Fix bug 2177 - encode DateTimeOffset value in generated skip token

* Fix bug 2177 - encode all types of parameters in generated skiptoken

Commit:
Avoid NullReferenceException on TruncatedCollection<T> (#2200)

* Add null check to TruncatedCollection.

* Add test.

Co-authored-by: Brecht Debaere <Brecht.Debaere@spray.com>

bdebaere committed:
Add cast operation to FilterBinder.BindIn, PR #2195

xuzhg committed:
Enable convention routing for operation import

gathogojr committed:
Fix Delta<TStructuralType>.GetInstance() method (#2246)

* Fix for complex type properties not being returned by Delta_Of_T GetInstance method

* Test to verify complex type and nested complex type are returned when DeltaOfT GetInstance method is called

Co-authored-by: John Gathogo <jogathogo@microsoft.com>

ificator committed:
Add HEAD Support for Attribute Routing (#2252)

* Add HEAD support

* Fix AspNetCore tests

gathogojr committed:
Fix invalid location header generated when key property for a model contains unicode chars (#2248)

* Fix invalid location header generated when key property for a model contains unicode chars

* Fix imports on test file

* Address escaping of unicode characters in @odata.editLink highlighted by user

* Fix failing test for Microsoft.AspNet.OData

* Harmonize assert statements in added tests

* Fix file alignment and address review feedback

Co-authored-by: John Gathogo <jogathogo@microsoft.com>

marabooy commented on src/Microsoft.AspNet.OData.Shared/Interfaces/IWebApiContext.cs:
```diff
@@ -1,12 +1,12 @@
 using Microsoft.AspNet.OData.Routing.Conventions;
 using Microsoft.OData.UriParser;
 using Microsoft.OData.UriParser.Aggregation;
+using System;
```
Sort namespaces

marabooy commented on src/Microsoft.AspNet.OData.Shared/Query/Validators/ODataQueryValidator.cs:
```diff
@@ -62,15 +64,22 @@ public virtual void Validate(ODataQueryOptions options, ODataValidationSettings
             }
 
-            if (options.Count != null || options.InternalRequest.IsCountRequest())
+            if (options is ODataQueryOptions _options)
```
Since we have moved to the Interface shouldn't we allow custom implementations of IODataQueryOptions to work as with this current implementation custom implementations are currently locked out.

kccarter76 replied:
InternalRequest does not exist on the interface because it is access is internal. on line 69, can || _options.InternalRequest.IsCountRequest() be factored out?

marabooy commented on src/Microsoft.AspNetCore.OData/Query/ODataQueryOptionsOfTEntity.cs:
```diff
@@ -2,16 +2,20 @@
 using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
 using Microsoft.OData.Edm;
+using System;
```
Sort namespaces so system comes first.

marabooy commented on test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/SUT/CollectionSerializerCustomersController.cs:
```diff
@@ -0,0 +1,44 @@
+﻿using Microsoft.AspNet.OData.Interfaces;
+using Microsoft.AspNet.OData.Query;
+using Microsoft.AspNet.OData.Test.Abstraction;
+using System.Linq;
```
Namespaces

marabooy commented:
@kccarter76 Thank you for your contribution, Kindly address the above comments.

kccarter76 committed:
update code files based on comments by marabooy

kccarter76 committed:
per maraboy removed the notsupported exception

bdebaere committed:
PR #2201: Modify ToInt16 to ToUInt16

marabooy committed:
PR #2167, Add fine grained cors access control support

xuzhg committed:
Update the ODL dependency to ODL 7.7.1 and release version to 7.5 (#2280)

* 1) Update the ODL dependency to ODL 7.7.1,
2) Update the release version to 7.5.0

* Fix the failing test cases

* Fix the failing test cases

* Fix the nested  context Url

* fix the failing e2e test cases

xuzhg committed:
Change the debugType as portable to push snupkg into Nuget.org

xuzhg committed:
Remove the null check, fix issue #1755 (#2142)

* Remove the null check, fix issue #1755

* Use the Data source provider to remove the null check for the EF5 and EF6

habbes committed:
Improve action selector routing (#2168)

* Make OdataActionSelector more strict when selection best candidate action

* Add tests for ODataActionSelector

* Fix code issues in OdataActionSelector

* Update ODataActionSelector fix for net core 3.0

* Port ODataActionSelector fix to .Net classic

* Fix ODataActionSelector for AspNet classic

* Minor code fixes in ODataActionSelectorTest

* Update ODataActionSelector Tests for .Net Core 3.1

* Fix ODataActionSelector tests build errors

* Fix some failing e2e tests

* Remove content headers from batch requests without body to fix routing.

* Minor refactoring of ODataActionSelector

* Add doc comments to ODataActionSelectorTestHelper

* Remove unnecessary content-length header from unrelated tests

* Restore files that shouldn't have been modified

* Restore files that shouldn't have been modified

* Use request method to determine whether request has a body or not

* Check action's declared methods in attributes to find best match if there are multiple candidates

* Minor refactor

* Ignore action parameters that are not model-bound from the request

* Fix e2e tests

* Add MERGE as possible http method for requests with body

* Prioritize actions with same number of parameters as keys

* Refactor ODataActionSelector

* Simply asp net classic ODataActionSelector

* Update tests to take account of key count odata route value

* Fix error in aspnet classic

* Update aspnet PublicApi.bsl

* Including navigation properties and function parameters in odata route key count

* Don't reject actions for missing navigationProperty param

* Move route keyCount to routing conventions store

* Routing conventions store tests compatible between aspnet and core

* Add doc comments to ODataActionSelector.TryMatch

* Address review comments

* Fix typo in comment

* Address PR comments

xuzhg committed:
Continue on the Nullcheck remove fix (#2285)

* Fix issue #2284 nullable check has regression on the EF 5/ 6

* Update the logic

KenitoInc committed:
Fix #2265: Child collection nextlink

gathogojr committed:
Implement support for long URL pattern (#2191)

* Implement support for long url patter

* Add newline and a dot

* Attempt to fix issue causing build to fail on ADO

* Fix broken tests

* Address review feedback

* Update PublicApi baseline

* Remove logic to resetting request stream position to 0

* Restore file from origin

* Reference shared test files the right way and add requested test

* Change Parse method in IODataQueryOptionsParser to ParseAsync. Fixes test failing intermittently

* Update baseline

* Add test requested in review

* Implement changes requested in PR review

* Remove empty ItemGroup from project file plus add helpful comments

* Update Public Api baseline

* Implement changes requested in review

* Use StreamReader in query options parser

* Implement changes requested in PR review

* Implement changes requested in PR review

* Add logic for identifying parser capable of parsing request options in request body to ODataQueryOptionsParserFactory

* Add test for adding custom query options parser

* Update PublicApi.bsl

* Implement improvement suggestion on PR and fix inadvertent dependency version conflict

* Missed 1 minor improvement :(

Co-authored-by: John Gathogo <jogathog@microsoft.com>
Co-authored-by: John Gathogo <jogathogo@microsoft.com>

xuzhg committed:
ResourceContext.ResourceInstance throws exception if there's a select or expand (#2288)

* Fix the ResourceContext.ResourceInstance throwing exception

* Update the Core version test

xuzhg committed:
Fix issue#2264, enable enum member alias for query options (#2267)

* Fix issue#2264, enable enum member alias for query options

* Resolve the comments

xuzhg committed:
Fix #2056: opertation returns path for the return entity set or singleton, PR #2060

xuzhg committed:
Add the () for the function call to fix the failing test cases

AdamCaviness committed:
Fix #2071: AspNetCore Resolving of Batch ContentId for AddRelatedObject.PR #2232

Co-authored-by: mikepizzo <mikep@microsoft.com>

KenitoInc committed:
Fix #2262: Invalid partial JSON response happens expanding an entity with more items than the allowed PageSize, PR #2287

Co-authored-by: mikepizzo <mikep@microsoft.com>

iwate committed:
Enhance batch accept header decision (#2231)

* add test cases

* modify CreateODataBatchResponseAsync logic

* fix typo

* Added handling for default q-value of 1.

* Added logic and tests to AspNet classic.

Co-authored-by: mikepizzo <mikep@microsoft.com>

Sreejithpin committed:
Add Instance annotations to WebAPI (#2219)

* InstanceAnnotationChanges

InstanceAnnotationChanges

* Update ODataResourceDeserializerTests.cs

* Add review comments

* Review comment fix

* Code review changes

Code review changes

* updates

* Update ODataResourceSerializer.cs

* changes after rebase

* InstanceAnnotationChanges

InstanceAnnotationChanges

* Add review comments

* Review comment fix

* Code review changes

Code review changes

* Update Microsoft.AspNet.OData.PublicApi.bsl

* public api

* public api change

* Address comments

* Changes

* Update ODataResourceSerializer.cs

* changes

* Comment changes

* Update ODataResourceSerializer.cs

* publi api

* comment chantes

* extra test

* Update GlobalSuppressions.cs

* name change

* changes

* changes and test

* Update ODataResourceValueSerializer.cs

* changes

* Update ODataResourceValueSerializer.cs

* changes

* changes

* updates

* Update ODataInstanceAnnotationContainer.cs

xuzhg committed:
Update version to 7.5.1

t03apt committed:
add support for parameters defined with type

gathogojr committed:
Fix buggy behaviour when addressing derived types (#2310)

* Fix buggy behaviour when addressing derived types

* Make improvements suggested in review

Co-authored-by: John Gathogo <jogathogo@microsoft.com>

gathogojr committed:
Fix failing tests occassioned by use of await expression in test asserts (#2314)

Co-authored-by: John Gathogo <jogathogo@microsoft.com>

ElizabethOkerio committed:
group by on date throws exceptions (#2302)

* group by on date throws exceptions

* fix based on review comments

ificator committed:
issue #2316, Select a complex with null value fails, PR #2320

habbes committed:
Issue #2301, Fix routing regression causing 404 on actions with [FromQuery], PR #2315

gathogojr committed:
Fix regression affecting usage of "in" expression with primitive collection property on RHS (#2323)

* Fix regression affecting usage of in expression with primitive collection property on rhs

* Implement review suggestions

* Use Theory to avoid repetition

Co-authored-by: John Gathogo <jogathogo@microsoft.com>

xuzhg committed:
Throw exception if the function call without providing the parameter … (#2306)

* Throw exception if the function call without providing the parameter value

* Make the ASP.NET Classic test pass

KenitoInc committed:
Issue #2255,Create the correct expression when casting entity/complex type to derived type. PR #2299

xuzhg committed:
compare Non-nullable guid

gathogojr committed:
Fix invalid next page link in nested collections (#2332)

* Fix invalid next page link in child collections

* Make variable local

* Added test for valid next page link generation

* Implement review suggestions

* Remove unnecessary else block

Co-authored-by: John Gathogo <jogathogo@microsoft.com>

mikepizzo committed:
Full async support (#2336)

* Full async support

* Wrap streams passed to input/output formatters to ensure only calling async methods

* Move from Wait() to async. May have to expose as new method for back compat.

* Improvements

* Elide trailing awaits

habbes committed:
Update version to 7.5.2 (#2355)

xuzhg committed:
Update ODL dependency to 7.7.3

xuzhg committed:
Enable JSON Metadata (#2361)

* Enable JSON Metadata

* Fix the failing test cases

* address the comment

KenitoInc committed:
Enable case insensitive request binding (#2311)

kccarter76 commented:
I do not have write access to the destination branch. the conflicts that are currently listed are a product of natural development and where not present at the time this was completed. I may not have the time available to address.

habbes committed:
Fix bug in attribute convention route values (#2368)

marabooy commented on test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/SUT/CollectionSerializerCustomersController.cs:
```diff
@@ -0,0 +1,43 @@
+    }
+}
+
+namespace Microsoft.AspNet.OData.Test.Formatter.Interface
```
Split this to two namespaces for consistency.

kccarter76 replied:
not sure about what you are asking here? been awhile since looking at this code.

kccarter76 replied:
I think you are asking about the sut class version in a class namespace and the interface implementation in an interface namespace. is that correct?

kccarter76 replied:
it might be best to split into two separate files to simplify reading the code base.

kccarter76 replied:
I need more eyes on this, since I have been unable to resolve the issue with the unit tests not running. @marabooy, I have sent you a collaboration request. I have hit a road block and the work crews appear to be on a permanent coffee break.

marabooy reviewed (commented):
Looks good to me, I didn't find any apparent issues with backwards compatibility. 
@kccarter76 Do let me know when you are ready to get this merged in and any help you require.

RamjotSingh committed:
Enabling query validation EnableQuery before action execution (#2362)

* Enabling query validation EnableQuery before action execution

* Adding test

* Name updates.

* Adding #if for NetCore tests

* Adding code to throw BadRequest

* Fixing issues

* Fixing the public API file

* Adding fixes for tests

* Fixing generic type code

* Adding support for ActionResult<>

* Taking PR comments

* Taking PR comments

* Taking PR comment

* Taking PR comments

* Just retrying build

Commit:
Fix duplicate batch Cookie header.

KenitoInc committed:
Create ODataError in Non-Success responses (#2341)

marabooy committed:
Fix behaviour for Preflight requests on the ODataBatchMiddleware. (#2356)

bratchikov committed:
Fix MergeIndividualAndBatchPreferences method (#2377)

* Fix MergeIndividualAndBatchPreferences method
for same Prefer headers case (bug with an unnecessary comma at the end of line)

* Add test merging same Prefer header while processing batch.

Co-authored-by: Tsar Nikolay <nsmirnov@ics.perm.ru>

xuzhg committed:
Update the release version to 7.5.4

xuzhg committed:
Update WebStack.versions.settings.targets

kccarter76 committed:
seperated the interface and class name space for this sut into seperate files.

kccarter76 commented:
to resolve the above conflicts, it would be best to collaborate. I am not sure what needs to be kept and merge what is needed to support the interface. EnableQueryAttribute was straight forward. however the others not so much.

RamjotSingh committed:
Moving EnableQuery request level query data storage to HttpContext (#2397)

* Moving request query data to HttpContext

* Adding one more E2E test

* Adding tests for parallelization check

marabooy commented:
@kccarter76 the BSL updates can be done by running the respective PublicApi tests and copying over the new file. 
Checkout the public api test linked [here](https://github.com/OData/WebApi/blob/5a5c585c139cddc3c63488f03daeb97a4b1ba868/test/UnitTest/Microsoft.AspNet.OData.Test.Shared/PublicApi/PublicApiTest.cs#L34)

kccarter76 commented:
I am missing code for RequestQueryData. I may need to merge/pull the current master into my github fork. any idea on how to do that with github?

kccarter76 committed:
Merge branch 'master' of https://github.com/OData/WebApi into OData-master

# Conflicts:
#	src/Microsoft.AspNetCore.OData/EnableQueryAttribute.cs
#	test/UnitTest/Microsoft.AspNet.OData.Test/PublicApi/Microsoft.AspNet.OData.PublicApi.bsl
#	test/UnitTest/Microsoft.AspNetCore.OData.Test/PublicApi/Microsoft.AspNetCore.OData.PublicApi.bsl
#	test/UnitTest/Microsoft.AspNetCore.OData.Test/PublicApi/Microsoft.AspNetCore3x.OData.PublicApi.bsl

kccarter76 committed:
resolved conflicts

kccarter76 commented:
is this being addressed in the AspNet solution:

Severity	Code	Description	Project	File	Line	Suppression State
Warning	CA0507	Post-build Code Analysis (FxCopCmd.exe) has been deprecated in favor of FxCop analyzers, which run during build. Refer to https://aka.ms/fxcopanalyzers to migrate to FxCop analyzers.				Active

kccarter76 committed:
no message

kccarter76 committed:
resolve conflct

marabooy commented on test/UnitTest/Microsoft.AspNet.OData.Test/PublicApi/Microsoft.AspNet.OData.PublicApi.bsl:
```diff
@@ -2354,9 +2354,48 @@ public enum Microsoft.AspNet.OData.Query.SelectExpandType : int {
 	Disabled = 2
 }
 
+<<<<<<< HEAD
```
Fix merge result

kccarter76 replied:
working on it, currently dealing with an issue where VS 2019 16.8 will not run unit tests on local machine. see tail end of microsoft/vstest#1220

BabaDorin commented:
Any updates?

drmcclelland commented:
Any updates @kccarter76 ?
